src/Access/AccessControl.cpp

```diff
@@ -559,33 +606,41 @@ AccessChangesNotifier & AccessControl::getChangesNotifier()
 }
 
 
-AuthResult AccessControl::authenticate(const Credentials & credentials, const Poco::Net::IPAddress & address) const
+AuthResult AccessControl::authenticate(const Credentials & credentials, const Poco::Net::IPAddress & address, const String & forwarded_address) const
 {
+    std::shared_ptr<const EnabledQuota> authentication_quota;
     try
```

Comment:
This function looks a bit too much complicated now. It's not necessary to check the quota twice (in `checkExceeded` and in `used`), the first `checkExceeded` should be enough. I believe this function can be written in a less complicated way:

```
auto authentication_quota = getAuthenticationQuota(credentials.getUserName());
if (authentication_quota)
    checkAuthenticationQuota(*authentication_quota, credentials.getUserName()); /// throws QUOTA_EXCEEDED with user_name if exceeded

AuthResult auth_result;
try
{
    auth_result = MultipleAccessStorage::authenticate(...);
}
catch (...)
{
    tryLogCurrentException();
    if (authentication_quota)
        authentication_quota->used(... /* check_exceeded */ false); /// already checked before
    throw ErrorCodes::AUTHENTICATION_FAILED
}

if (authentication_quota)
    authentication_quota->reset(QuotaType::FAILED_SEQUENTIAL_AUTHENTICATIONS);

return auth_result;
```

Reply:
Thanks for the suggestion!

Let's imagine the situation `without` checking `quota->used(throw_exception = true)`:
    We have a quota with FAILED_SEQUENTIAL_AUTHENTICATIONS = 1.
The user tries to login with the wrong password the first time:
      * Call `check_exceed` (0 > 1 - FALSE)
      * throw `AUTHENTICATION_FAILED` error (`used_tries` becomes 1) 
The user tries to login with the wrong password the second time: 
      * Call check_exceed (1 > 1 - FALSE)
      * throw `AUTHENTICATION_FAILED` error from  `throw ErrorCodes::AUTHENTICATION_FAILED` (`used_tries` becomes 2) // <<-- This looks like the wrong behavior for me. I would desire to see the 'QUOTA_EXCEED' error here.
      
So the line:
```cpp
  if (authentication_quota)
        authentication_quota->used(... /* check_exceeded */ false);
```

It looks more correct if `check_exceeded` will be `true`. Then the sequence will be:

The user tries to login with the wrong password the first time:
      * Call `check_exceed` (0 > 1 - FALSE)
      * throw `AUTHENTICATION_FAILED` error (`used_tries` becomes 1) 
The user tries to login with the wrong password the second time:
      * Call check_exceed (1 > 1 - FALSE)
      * throw `QUOTA_EXCEED` from `authentication_quota->used(... /* check_exceeded */ true);` (`used_tries` becomes 2)
      

Reply:
Though it looks a bit strange that with `FAILED_SEQUENTIAL_AUTHENTICATIONS = 1` we allow two login attempts. And there is no way to allow only one login attempt.

Reply:
I don't like it either, but it's a feature of quotas where the condition is violated when `used > max.`

In this specific case, the condition `used >= max` would look better from the user's perspective. I would change a code inside `EnabledQuota.cpp` in these lines:
https://github.com/ClickHouse/ClickHouse/blob/dfc761ce5f4ffd360c72c37a53ffe27eebe5d86f/src/Access/EnabledQuota.cpp#L56-L66
https://github.com/ClickHouse/ClickHouse/blob/dfc761ce5f4ffd360c72c37a53ffe27eebe5d86f/src/Access/EnabledQuota.cpp#L83-L89

Instead we can use different conditions depending on `QuotaType quota_type` like:

```cpp
static bool compareQuotaValue(QuotaValue used, QuotaValue max, QuotaType quota_type)
{
    return quota_type == QuotaType::FAILED_SEQUENTIAL_AUTHENTICATIONS ? used >= max : used > max;
}
```

and use `CompareQuotaValue(used, max, quota_type)` instead of just `used > max`. 
But I need the confirmation that making such change satisfies ClickHouse.

Reply:
Now the code works as you proposed but with `authentication_quota->used(... /* check_exceeded */ true);`

The default-generated exception from `authentication_quota->checkExceed(...)` already contains the `user name`:

```
Code: 201. DB::Exception: Received from localhost:9000. DB::Exception: Quota for user `2884_user_579116` for 3155695200s has been exceeded: failed_sequential_authentications = 2/1. Interval will end at 2069-12-31 06:00:00. Name of quota template: `2884_quota_579116`. (QUOTA_EXCEEDED)
```

Reply:
> and use CompareQuotaValue(used, max, quota_type) instead of just used > max.
But I need the confirmation that making such change satisfies ClickHouse.

I have another idea: let's just remove the call `authentication_quota->used()` after `catch (...)` and call it at the beginning instead of `checkExceeded()`:

```
auto authentication_quota = getAuthenticationQuota(credentials.getUserName());
if (authentication_quota)
    authentication_quota->used(FAILED_SEQUENTIAL_AUTHENTICATIONS, 1);

try
{
    auto auth_result = MultipleAccessStorage::authenticate(...);

    if (authentication_quota)
        authentication_quota->reset(QuotaType::FAILED_SEQUENTIAL_AUTHENTICATIONS);

    return auth_result;
}
catch (...)
{
    tryLogCurrentException();
    throw ErrorCodes::AUTHENTICATION_FAILED
}
```

This way is not very intuitive however it seems it will work exactly as we want it to work. And it's also shorter.

We need a comment before `used()` though with a detailed description. About that we increase the counter of authentication failures in the beginning and reset it after a successful authentication. And we do that because if we don't have quota for a failed authentication then we shouldn't try to authenticate at all.

Reply:
The implementation variant you suggested will work with this feature:
`authentication_quota->used` will increase the failed authentication counter even if the quota is exceeded; that looks fine for me.

Reply:
Code is updated, 
Now it works as desired.